tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java

```diff
@@ -582,10 +593,29 @@ private static Map<TopicPartition, List<Integer>> calculateAssignment(Map<TopicP
         Map<TopicPartition, List<Integer>> proposedAssignments = new HashMap<>();
         groupedByTopic.forEach((topic, assignment) -> {
             List<Integer> replicas = assignment.get(0).getValue();
-            Map<Integer, List<Integer>> assignedReplicas = AdminUtils.
-                assignReplicasToBrokers(brokerMetadatas, assignment.size(), replicas.size());
-            assignedReplicas.forEach((partition, replicas0) ->
-                proposedAssignments.put(new TopicPartition(topic, partition), replicas0));
+            int partitionNum = assignment.size();
+            // generate topic assignments
+            TopicAssignment topicAssignment = REPLICA_PLACER.place(
+                    new PlacementSpec(0, partitionNum, (short) replicas.size()),
+                    new ClusterDescriber() {
+                        @Override
+                        public Iterator<UsableBroker> usableBrokers() {
+                            return brokerMetadatas.stream().map(brokerMetadata -> new UsableBroker(
+                                    brokerMetadata.id,
+                                    brokerMetadata.rack,
+                                    false
+                            )).iterator();
+                        }
+
+                        @Override
+                        public Uuid defaultDir(int brokerId) {
+                            return DirectoryId.MIGRATING;
+                        }
+                    });
+
+            for (PartitionAssignment partitionAssignment : topicAssignment.assignments()) {
+                proposedAssignments.put(new TopicPartition(topic, partitionNum), partitionAssignment.replicas());
```

Comment:
`partitionNum` is the count of partition number rather than partition index...
```
            for (int i = 0; i < topicAssignment.assignments().size(); i++) {
                proposedAssignments.put(new TopicPartition(topic, i), topicAssignment.assignments().get(i).replicas());
            }
```